python/ray/data/_internal/batcher.py

```diff
@@ -77,4 +122,191 @@ def next_batch(self) -> Block:
+        """Indicate to the batcher that no more blocks will be added to the batcher.
+
+        No more blocks should be added to the batcher after calling this.
+        """
+        self._done_adding = True
+
+    def has_any(self) -> bool:
+        """Whether this batcher has any data."""
+        return self._buffer_size() > 0
+
+    def has_batch(self) -> bool:
+        """Whether this batcher has any batches."""
+        buffer_size = self._buffer_size()
+        # If still adding blocks, ensure that removing a batch wouldn't cause the
+        # shuffle buffer to dip beneath its configured minimum size.
+        return buffer_size - self._batch_size >= self._buffer_min_size or (
+            self._done_adding and buffer_size >= self._batch_size
+        )
+
+    def _buffer_size(self) -> int:
+        """Return shuffle buffer size."""
+        buffer_size = self._builder.num_rows()
+        if self._shuffle_buffer is not None:
+            # Include the size of the concrete (materialized) shuffle buffer, adjusting
+            # for the batch head position, which also serves as a counter of the number
+            # of already-yielded rows from the current concrete shuffle buffer.
+            buffer_size += (
+                BlockAccessor.for_block(self._shuffle_buffer).num_rows()
+                - self._batch_head
+            )
+        return buffer_size
+
+    def next_batch(self) -> Block:
+        """Get the next shuffled batch from the shuffle buffer.
+
+        Returns:
+            A batch represented as a Block.
+        """
+        assert self.has_batch() or (self._done_adding and self.has_any())
+        # Add rows in the builder to the shuffle buffer.
+        if self._builder.num_rows() > 0:
+            if self._shuffle_buffer is not None:
+                if self._batch_head > 0:
+                    # Compact the materialized shuffle buffer.
+                    # TODO(Clark): If alternating between adding blocks and fetching
+                    # shuffled batches, this aggressive compaction could be inefficient.
+                    self._shuffle_buffer = BlockAccessor.for_block(
+                        self._shuffle_buffer
+                    ).take(self._shuffle_indices[self._batch_head :])
+                # Add the unyielded rows from the existing shuffle buffer.
+                self._builder.add_block(self._shuffle_buffer)
+            # Build the new shuffle buffer.
+            self._shuffle_buffer = self._builder.build()
+            # Reset the builder.
+            self._builder = DelegatingBlockBuilder()
+            # Invalidate the shuffle indices.
+            self._shuffle_indices = None
+            self._batch_head = 0
+
+        assert self._shuffle_buffer is not None
+        buffer_size = BlockAccessor.for_block(self._shuffle_buffer).num_rows()
+        # Truncate the batch to the buffer size, if necessary.
+        batch_size = min(self._batch_size, buffer_size)
```

Comment:
So looks we do continue to yield batches when the num of rows in the buffer drop below shuffle_buffer_min_size. Can you adjust the parameter documentation above to reflect this?

Reply:
Same as above.